### PR TITLE
[V3] Upgrade ACC-based encryption + JWT signing

### PR DESCRIPTION
**Pull request by hwrdtm**
# What

Closes LIT-804.
Closes LIT-695.
Related to LIT-951.

⚠️ **This is a massive PR. Brace yourself with some 🍿  and 🍵 !!** ⚠️ 

This PR:
- Deserves a major version bump. We are completely overhauling the ACC-based encryption and JWT signing schemes with new schemes, and are bringing in a number of incompatible changes.
- **ID-based encryption**: This new scheme is also known as Timelock Encryption, or TimeCrypt, where there exists an Identity Parameter, and the signature over it is a decryption key over a ciphertext that has been generated with this Identity Parameter. In other words, signature shares - with the Identity Parameter as the signature payload - are combined into a signature, and this signature can be used to decrypt an associated ciphertext. Read more in the spec here: https://www.notion.so/litprotocol/Switch-from-Signcryption-to-BLS-signature-is-key-scheme-b75c21927ed64511b523c8a93b870a65. 
  - Note that the encryption step is non-interactive and happens entirely client-side.
- **ACC-based JWT Signing (V2)**: This new scheme simplifies the overall flow by baking into each dapp endpoint the required access control conditions directly. This means that there is no "store signing condition" step as dapp developers just write it into their dapp code and an additional check against the access control condition is performed there. When clients request for a signed JWT with claims of the specified access control condition, the signature of this JWT is only provided when the nodes determine that the user satisfy this access control condition. Read more in the spec here: https://www.notion.so/litprotocol/ACC-based-JWT-signing-V2-a4db1446bac54eea90b7ab08e5455246.
- The BLS WASM (`bls-sdk.ts`) is entirely replaced with new code. Unit tests are added to this package.
- **The SDK no longer supports the `jalapeno` network moving forward. Developers can use the old SDK versions to continue using the old encryption and JWT signing schemes**. `serrano` network is the new default.
- Updated `crypto` package to be a lightweight wrapper over `bls-sdk` package with minor conversion logic.
- Updated `encryption` package to replace the old encryption + JWT signing logic with the new schemes. This is becoming more a helper library over `lit-node-client-nodejs` and `crypto`.
- Refactored `params-validators` to use a plugin-based approach with various `*Validator`s.
- Updated all manual tests in `apps/html` to use the new ID-based encryption and JWT signing schemes. You will notice quite a few test cases are removed since we no longer need to store encryption keys, or store signing conditions respectively.
- `demo-encrypt-decrypt-react`:
  - Turned this into a TS project.
  - Updated this demo to use the new ID-based encryption scheme - just a `encryptString` and `decryptToString` call.
- Implement `demo-locked-express-app` to demo the new JWT signing flow. Instructions in its `README.md`.
- Made use of generics in a number of places, eg. `IEither`, `SuccessNodePromises` and `benchmark()`

# Testing

- A large number of HTML manual tests have been tested (against local network)
- E2E tests pass
- Unit tests pass

# TODO

- [x] Re-test `demo-encrypt-decrypt-react` after rebasing
- [x] Re-test `demo-locked-express-app` after rebasing
- [ ] Update `demo-encrypt-decrypt-react/package.json` to use newly published SDK version.
- [ ] Update `demo-locked-express-app/package.json` to use newly published SDK version.
- [x] Remove support for `serrano` too and introduce `cayenne` as the new default.

### CODE REVIEW

File: packages/bls-sdk/src/lib/bls-sdk.d.ts [hunk 1]
```diff
@@ -0,0 +1,47 @@
+/**
```
Comment (hwrdtm) [inline packages/bls-sdk/src/lib/bls-sdk.d.ts:1]: the types don't seem to be linking properly...

File: cypress/e2e/spec.cy.ts [hunk 2]
```diff
@@ -62,255 +63,185 @@ describe('Encrypt and Decrypt String', () => {
...
-  });
-
-  it('saves encryption key', async () => {
-    // const { encryptedString, symmetricKey } = savedParams;
-
-    const encryptedSymmetricKey =
-      await savedParams.litNodeClient.saveEncryptionKey({
+describe('Encrypt and Decrypt', () => {
+  it('encrypts', async () => {
+    cy.window().then(async () => {
```
Comment (hwrdtm) [inline cypress/e2e/spec.cy.ts:82]: I'm not sure why, but without wrapping each functionality with this `cy.window()` the `wasm` is undefined and not initialized

File: apps/demo-encrypt-decrypt-react/package.json [hunk 3]
```diff
@@ -4,13 +4,18 @@
   "private": true,
   "dependencies": {
     "@lit-protocol/lit-node-client": "^2.1.158",
```
Comment (hwrdtm) [inline apps/demo-encrypt-decrypt-react/package.json:6]: TODO: need to update version once published.

File: apps/demo-locked-express-app/package.json [hunk 4]
```diff
@@ -0,0 +1,28 @@
+{
+  "name": "y",
+  "version": "1.0.0",
+  "main": "index.js",
+  "license": "MIT",
+  "dependencies": {
+    "@lit-protocol/lit-node-client-nodejs": "file:../../packages/lit-node-client-nodejs",
```
Comment (hwrdtm) [inline apps/demo-locked-express-app/package.json:7]: TODO: need to update version once published.

**Comment by hwrdtm** [comment]
Not sure why CI is failing, can't repro locally too with `npx nx run auth-helpers:test`

File: apps/demo-locked-express-app/index.ts [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+import * as LitJsSdk from "@lit-protocol/lit-node-client-nodejs";
+import cookieParser from "cookie-parser";
+import { MultipleAccessControlConditions } from '@lit-protocol/types';
+
+const app: Express = express();
+const port = process.env.PORT || 3001;
+
+const MY_OWN_WALLET_ADDRESS = "0x4259E44670053491E7b4FE4A120C70be1eAD646b";
+
+const accessControlCondtionsForProtectedPath1: MultipleAccessControlConditions = {
```
Comment (Ansonhkg) [inline apps/demo-locked-express-app/index.ts:12]: https://www.notion.so/ACC-based-JWT-signing-V2-a4db1446bac54eea90b7ab08e5455246?d=6f3ddd0b3bcb437c941f7ac1b89d4203&pvs=4#c97e0569a910425cb2e4eed7ed02500e

File: packages/auth-browser/src/lib/chains/sol.ts [hunk 6]
```diff
@@ -105,9 +106,11 @@ export const checkAndSignSolAuthMessage = async (): Promise<AuthSig> => {
 
   try {
     // when it's not in local storage, it's a string
+    // @ts-ignore
```
Comment (Ansonhkg) [inline packages/auth-browser/src/lib/chains/sol.ts:109]: We probably want to refactor the IEither type at some point so we don't have to do this

File: apps/demo-locked-express-app/index.ts [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+    console.log("JWT payload", payload);
+
+    // Verify the access control conditions in the JWT claims are as expected.
+    const expectedAccessControlConditionsHash = (await litNodeClient.getHashedAccessControlConditions(accessControlCondtionsForProtectedPath1))!.toString();
+    const actualAccessControlConditionsHash = (await litNodeClient.getHashedAccessControlConditions(payload))!.toString();
+    if (expectedAccessControlConditionsHash !== actualAccessControlConditionsHash) {
+      console.error("Access control conditions in JWT are not as expected");
+      res.status(401).send("Unauthorized");
+      return;
+    }
```
[the remark this answers is not in the document]
Comment (hwrdtm) [reply apps/demo-locked-express-app/index.ts:63]: Totally, though for demo purposes we don't have to be too pedantic

File: apps/demo-encrypt-decrypt-react/src/App.tsx [hunk 8]
```diff
@@ -0,0 +1,196 @@
...
+    code = code.replace('// { Loading... }', `// [string] { ${encryptRes.result.ciphertext} }`);
+    code = code.replace('// { Loading... }', `// [Uint8Array] { ${JSON.stringify(encryptRes.result.dataToEncryptHash)} }`);
+    setData(code);
+
+    // --------- NEXT STEP ---------
+    const decryptRes = await benchmark(async () => {
+      return LitJsSdk.decryptToString({
+        accessControlConditions: accs,
+        ciphertext: encryptRes.result.ciphertext,
+        dataToEncryptHash: encryptRes.result.dataToEncryptHash,
```
[the remark this answers is not in the document]
Comment (hwrdtm) [reply apps/demo-encrypt-decrypt-react/src/App.tsx:137]: Note: We can make the high level wrapper method to use a single struct.

**Review by glitch003 — APPROVED** [review]

File: apps/demo-locked-express-app/index.ts (hunk 5 above)
Comment (hwrdtm) [reply apps/demo-locked-express-app/index.ts:12]: Responded.

File: packages/bls-sdk/src/lib/bls-sdk.spec.ts [hunk 9]
```diff
@@ -6,10 +6,148 @@ global.TextDecoder = TextDecoder;
...
+    const ciphertext =
+      'l9a/01WDJB/euKxtbWcuQ8ez/c9eZ+jQryTHZVLN0kfd7XHoLs6FeWUVmk89ovQGkQJnnFDKjq6kgJxvIIrxXd9DaGuRBozLdA1G9Nk413YhTEqsENuHU0nSa4i6F912KltE15sbWKpDfPnZF6CA2UKBAw==';
+    const signatureShares = [
+      '01b2b44a0bf7184f19efacad98e213818edd3f8909dd798129ef169b877d68d77ba630005609f48b80203717d82092a45b06a9de0e61a97b2672b38b31f9ae43e64383d0375a51c75db8972613cc6b099b95c189fd8549ed973ee94b08749f4cac',
+      '02a8343d5602f523286c4c59356fdcfc51953290495d98cb91a56b59bd1a837ea969cc521382164e85787128ce7f944de303d8e0b5fc4becede0c894bec1adc490fdc133939cca70fb3f504b9bf7b156527b681d9f0619828cd8050c819e46fdb1',
+      '03b1594ab0cb56f47437b3720dc181661481ca0e36078b79c9a4acc50042f076bf66b68fbd12a1d55021a668555f0eed0a08dfe74455f557b30f1a9c32435a81479ca8843f5b74b176a8d10c5845a84213441eaaaf2ba57e32581584393541c5aa',
+    ];
+
+    // execute
+    const plaintext = blsSdk.verify_and_decrypt_with_signature_shares(
```
Comment (sarahzdev) [inline packages/bls-sdk/src/lib/bls-sdk.spec.ts:82]: (can't seem to comment on the `packages/bls-sdk/src/lib/bls-sdk.ts` file directly)

maybe we should strive to keep casing consistent across our sdk

**Comment by joshLong145** [comment]
Closing as functionality has been merged into: https://github.com/LIT-Protocol/js-sdk/tree/feat/SDK-V3

**Pull request closed without merging**